### ISSUE TITLE
Next generation rendering API in TypeScript

### ISSUE THREAD

**Issue by schani** [quicktype/quicktype#199]
The main things that need to be exposed are:

* Some form of the IR.  We should consider that in the future we'll want the IR to have providence information, i.e. which parts of the input each type was inferred from.

* Producing text.  We might want to support multiple files right away (we already need it for Java).  We should also consider that we'll want to preserve providence information in the output, so that we'll eventually be able to tell which parts of the output correspond to which parts of the input, and vice versa.

* Reporting issues/warnings.  These should also have location information, in the input and/or the output.

* An API for naming things.  I'd like to do this properly right away.  There are a few GitHub issues where I've written on this topic.

**Comment by schani** [comment]
Proposal for the representation of rendered documents (I'm using PureScript syntax for clarity):

```
data Document
    = Text String -- without newlines
    | Newline
    | Name Name
    | Array Document
    | Annotated Annotation Document
```

I think I want to separate newlines out because it will make counting lines easier.  I'd rather have positions given as (line, column) rather than index into a string, if only because the latter is sure to lead to severe unicode trouble (if we're off by a few characters in a single line it's not so bad, but if it's until the end of the document that's a big issue).

The interesting bits here are the annotations and the names.  For now I know two purposes for the annotations:

* To carry provenance information.
* To indicate issues/warnings.

These two can be intermixed.  In fact, since annotations can be nested, it's possible to have nested provenances (also nested issues, whether that makes sense or not).  For example, the whole text of a `class` declaration would point to the JSON object (or multiple JSON objects) it was inferred from, whereas a single `int` declaration within that class would point to the integer in the JSON that `int` came from.

Representing names as first-class entities in the document, as opposed to just pasting them in as strings, gives us the ability to allow users to interactively change names in the result document.  If we keep the name graph around (see my comment on #102) we can also still fix collisions.  We could even show users which names were conflicting, leading to renaming.

**Comment by schani** [comment]
Merged in bd06db2ae6e0d99d5aa6b02ce96eacd9752b2ac1
